src/platform_impl/linux/x11/event_processor.rs

```diff
@@ -1221,30 +1227,29 @@ impl<T: 'static> EventProcessor<T> {
     }
 
     fn handle_pressed_keys<F>(
-        &self,
+        wt: &super::EventLoopWindowTarget<T>,
         window_id: crate::window::WindowId,
         state: ElementState,
+        mod_keymap: &ModifierKeymap,
+        device_mod_state: &mut ModifierKeyState,
         callback: &mut F,
     ) where
         F: FnMut(Event<'_, T>),
     {
-        let wt = get_xtarget(&self.target);
-
         let device_id = mkdid(util::VIRTUAL_CORE_KEYBOARD);
-        let modifiers = self.device_mod_state.modifiers();
-
-        // Get the set of keys currently pressed and apply Key events to each
-        let keys = wt.xconn.query_keymap();
-
-        for keycode in &keys {
-            if keycode < 8 {
-                continue;
-            }
+        let modifiers = device_mod_state.modifiers();
 
+        // Get the set of keys currently pressed and apply Key events to each and also updating
+        // the pressed modifiers.
+        for keycode in wt.xconn.query_keymap().into_iter().filter(|k| *k >= 8) {
             let scancode = (keycode - 8) as u32;
             let keysym = wt.xconn.keycode_to_keysym(keycode);
             let virtual_keycode = events::keysym_to_element(keysym as c_uint);
 
+            if let Some(modifier) = mod_keymap.get_modifier(keycode as ffi::KeyCode) {
+                device_mod_state.key_event(state, keycode, modifier);
+            }
```

Comment:
Without looking at the rest of the code, I'd assume this is the place where the bugfix actually happens and the modifiers state tracking is fixed? So `key_event` function on the `device_mod_state` is responsible for updating the modifiers?

Reply:
Correct.

Without this the modifier state change needed from the synthetic key events triggered by `FocusIn` never happened.